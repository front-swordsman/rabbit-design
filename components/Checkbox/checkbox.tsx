import * as React from 'react'
import { CheckboxItemProps, CheckboxProps } from '../../types/checkbox'
import CheckboxItem from './checkboxItem'
import { noop } from '../utils/tool'

const prefixCls = 'snake-checkbox'

const Checkbox: React.SFC<CheckboxProps> & {
  item: React.ForwardRefExoticComponent<CheckboxItemProps & React.RefAttributes<{}>>
} = ({ onChange = noop, disabled = false, options = [], value = [] }) => {
  const handleChange = (checked: boolean, checkedValue: string | number) => {
    const cloneValue = value.slice()
    if (checked) {
      cloneValue.push(checkedValue)
    } else {
      cloneValue.splice(cloneValue.indexOf(checkedValue), 1)
    }
    onChange(cloneValue)
  }

  return (
    <div className={`${prefixCls}`}>
      {options.map(p => {
        const checked = value.includes(p.value)
        return (
          <CheckboxItem
            key={p.value}
            checked={checked}
            disabled={disabled || p.disabled}
            autoFocus={p.autoFocus}
            onChange={checked => handleChange(checked, p.value)}
          >
            {p.label}
          </CheckboxItem>
        )
      })}
    </div>
  )
}

Checkbox.item = CheckboxItem

export default Checkbox
