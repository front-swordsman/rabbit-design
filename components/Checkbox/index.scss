@import '../styles/themes/default';

$checkbox: 'snake-checkbox-item';

.#{$checkbox} {
  font-size: $hd * 14;
  font-variant: tabular-nums;
  line-height: $lineHeight;
  color: rgba(0, 0, 0, 0.65);
  box-sizing: border-box;
  margin: 0;
  padding: 0;
  list-style: none;
  cursor: pointer;
  display: inline-block;

  &:hover {
    .#{$checkbox}-select-inner {
      border-color: $brand;
    }
  }

  &-select {
    font-size: $hd * 14;
    font-variant: tabular-nums;
    color: rgba(0, 0, 0, 0.65);
    box-sizing: border-box;
    margin: 0;
    padding: 0;
    list-style: none;
    white-space: nowrap;
    cursor: pointer;
    outline: none;
    display: inline-block;
    line-height: $lineHeight-1;
    position: relative;
    vertical-align: middle;
    top: -0.09em;

    input {
      position: absolute;
      left: 0;
      z-index: 1;
      cursor: pointer;
      opacity: 0;
      top: 0;
      bottom: 0;
      right: 0;
      width: 100%;
      height: 100%;
    }

    &-inner {
      position: relative;
      top: 0;
      left: 0;
      display: block;
      width: 16px;
      height: 16px;
      border: 1px solid #d9d9d9;
      border-radius: 2px;
      background-color: #fff;
      transition: all 0.3s;
      border-collapse: separate;

      &::after {
        transform: rotate(45deg) scale(0);
        position: absolute;
        left: 4.57142857px;
        top: 1.14285714px;
        display: table;
        width: 5.71428571px;
        height: 9.14285714px;
        border: 2px solid #fff;
        border-top: 0;
        border-left: 0;
        content: ' ';
        transition: 0.1s $easeInBack, opacity 0.1s;
        opacity: 0;
      }
    }
  }

  &-label {
    padding: 0 8px;
  }

  &-checked {
    .#{$checkbox}-select-inner {
      background-color: $brand;
      border-color: $brand;

      &::after {
        transform: rotate(45deg) scale(1);
        position: absolute;
        display: table;
        border: 2px solid #fff;
        border-top: 0;
        border-left: 0;
        content: ' ';
        transition: all 0.2s $easeOutBack 0.1s;
        opacity: 1;
      }
    }

    .#{$checkbox}-select {
      &::after {
        position: absolute;
        top: 0;
        left: 0;
        width: 100%;
        height: 100%;
        border-radius: 2px;
        border: 1px solid $brand;
        content: '';
        animation: antCheckboxEffect 0.36s ease-in-out;
        animation-fill-mode: both;
        visibility: hidden;
      }
    }
  }

  &-indeterminate {
    .#{$checkbox}-select-inner {
      background-color: #fff;
      border-color: #d9d9d9;

      &::after {
        content: ' ';
        transform: translate(-50%, -50%) scale(1);
        border: 0;
        left: 50%;
        top: 50%;
        width: 8px;
        height: 8px;
        background-color: $brand;
        opacity: 1;
      }
    }
  }

  &-disabled {
    .#{$checkbox}-select {
      cursor: not-allowed;
    }

    .#{$checkbox}-select-inner {
      border-color: #d9d9d9 !important;
      background-color: #f5f5f5;

      &::after {
        animation-name: none;
        border-color: #f5f5f5;
        border-collapse: separate;
      }
    }

    &.#{$checkbox}-checked {
      .#{$checkbox}-select-inner {
        &::after {
          animation-name: none;
          border-color: rgba(0, 0, 0, 0.25);
        }
      }
    }
  }

  input[type='checkbox'][disabled] {
    cursor: not-allowed;
  }
}
