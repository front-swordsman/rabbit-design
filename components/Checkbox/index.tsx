import CheckboxGroup from './checkboxGroup'

export default CheckboxGroup
