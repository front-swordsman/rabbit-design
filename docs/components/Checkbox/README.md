---
imports:
  import CheckboxDemoCode from '!raw-loader!./demo.tsx';
  import CheckboxDemo from './demo.tsx';
  import Block from '../../layout/block';
---

# Checkbox 复选框

在一组可选项中进行多项选择时

## 简单示例

:::demo
<Block code={CheckboxDemoCode} des="复选框">
  <CheckboxDemo />
</Block>
:::

## Api

### Checkbox

| 参数     | 说明              | 类型   | 可选值 | 默认值 |
| -------- | ----------------- | ------ | ------ | ------ |
| checked | 是否选中 | boolean | -- | `false`
| onChange | 触发事件的回调 | function(checked: boolean) : void | -- | noop
| autoFocus | 是否自动聚焦 | boolean | -- | `false`
| disabled | 是都不可选 | boolean | -- | `false`
| className | 类名 | string | -- | --
| indeterminate | 是否半选 | boolean | -- | `false`

### CheckboxGroup

| 参数     | 说明              | 类型   | 可选值 | 默认值 |
| -------- | ----------------- | ------ | ------ | ------ |
| options | 数据源 | Data[] | -- | []
| disabled | 是否全部不可用 | boolean | -- | `false`
| value | 选中的值的集合 | string[] | -- | []
| onChange | 触发事件的回调 | function(value: string[]) : void | -- | --

### Data

| 参数     | 说明              | 类型   | 可选值 | 默认值 |
| -------- | ----------------- | ------ | ------ | ------ |
| label | 文本信息 | ReactNode | 否 | --
| value | 用于选定的值 | string | 否 | --
| disabled | 是否不可用 | boolean | -- | --
| autoFocus | 是否自动聚焦 | boolean | -- | --