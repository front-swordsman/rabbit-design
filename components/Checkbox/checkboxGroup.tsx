import * as React from 'react'
import { CheckGroupProps, CheckboxProps } from './../../types/checkbox'
import Checkbox from './checkbox'
import { noop } from '../utils/tool'

const prefixCls = 'snake-checkbox'

const CheckboxGroup: React.SFC<CheckGroupProps> & {
  checkbox: React.ForwardRefExoticComponent<CheckboxProps & React.RefAttributes<{}>>
} = ({ onChange = noop, disabled = false, options = [], value = [] }) => {
  const handleChange = (checked: boolean, checkedValue: string) => {
    const cloneValue = value.slice()
    if (checked) {
      cloneValue.push(checkedValue)
    } else {
      cloneValue.splice(cloneValue.indexOf(checkedValue), 1)
    }
    onChange(cloneValue)
  }

  return (
    <div className={`${prefixCls}-group`}>
      {options.map(p => {
        const checked = value.includes(p.value)
        return (
          <Checkbox
            key={p.value}
            checked={checked}
            disabled={disabled || p.disabled}
            autoFocus={p.autoFocus}
            onChange={checked => handleChange(checked, p.value)}
          >
            {p.label}
          </Checkbox>
        )
      })}
    </div>
  )
}

CheckboxGroup.checkbox = Checkbox

export default CheckboxGroup
