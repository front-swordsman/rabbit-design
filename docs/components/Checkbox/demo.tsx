import * as React from 'react'
import { Checkbox } from 'components'

const CheckboxItem = Checkbox.item
const options = [
  {
    label: '篮球',
    value: 'basketball'
  },
  {
    label: '足球',
    value: 'football'
  },
  {
    label: '滑板',
    value: 'skateboard'
  },
  {
    label: '跳伞',
    value: 'parachute'
  }
]

export default function CheckboxDemo() {
  const [checked, setChecked] = React.useState(false)
  const [value, setValue] = React.useState([])
  console.log(value)
  return (
    <div>
      <h3>基础用法</h3>
      <CheckboxItem checked={checked} onChange={checked => setChecked(checked)}>
        点击选择
      </CheckboxItem>
      <h3>不可用</h3>
      <CheckboxItem checked disabled>
        点击选择
      </CheckboxItem>
      <Checkbox disabled>点击选择</Checkbox>
      <h3>group 用法</h3>
      <div>
        <CheckboxItem indeterminate={value.length && value.length < 4} checked={value.length === 4}>
          全部
        </CheckboxItem>
      </div>
      <Checkbox options={options} value={value} onChange={value => setValue(value)} />
    </div>
  )
}
