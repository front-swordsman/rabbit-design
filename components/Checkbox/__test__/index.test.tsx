import * as React from 'react'
import { render, fireEvent } from 'react-testing-library'
import Checkbox from '../index'

describe('checkbox test', () => {
  const CheckboxItem = Checkbox.item
  const options = [
    {
      label: '篮球',
      value: 'basketball'
    },
    {
      label: '足球',
      value: 'football'
    },
    {
      label: '滑板',
      value: 'skateboard'
    },
    {
      label: '跳伞',
      value: 'parachute'
    }
  ]

  it('snapshot', () => {
    const checked = render(<CheckboxItem checked>选中</CheckboxItem>)
    const notChecked = render(<CheckboxItem>未选中</CheckboxItem>)
    const disabled = render(<CheckboxItem disabled>不可用</CheckboxItem>)
    const indent = render(<CheckboxItem indeterminate>半选</CheckboxItem>)
    expect(checked.container.innerHTML).toMatchSnapshot()
    expect(notChecked.container.innerHTML).toMatchSnapshot()
    expect(disabled.container.innerHTML).toMatchSnapshot()
    expect(indent.container.innerHTML).toMatchSnapshot()
  })

  it('fire change', () => {
    function CheckDemo() {
      const [checked, setChecked] = React.useState(false)
      return (
        <CheckboxItem checked={checked} onChange={checked => setChecked(checked)}>
          测试demo
        </CheckboxItem>
      )
    }
    const { getByText } = render(<CheckDemo />)
    const inputNode = getByText(/测试demo/i).previousSibling.firstChild as HTMLInputElement
    expect(inputNode.checked).toBe(false)
    fireEvent.click(inputNode)
    expect(inputNode.checked).toBe(true)
  })

  it('checkbox group', () => {
    const { container } = render(<Checkbox options={options} />)
    expect(container.firstChild.childNodes.length).toBe(4)
  })

  it('fire group change', () => {
    function GroupDemo() {
      const [value, setValue] = React.useState([])
      return <Checkbox value={value} onChange={value => setValue(value)} options={options} />
    }
    const { getByText } = render(<GroupDemo />)
    const baskNode = getByText(/篮球/i).previousSibling.firstChild as HTMLInputElement
    fireEvent.click(baskNode)
    expect(baskNode.checked).toBe(true)
    fireEvent.click(baskNode)
    expect(baskNode.checked).toBe(false)
  })

  it('disabled checkbox', () => {
    const fn = jest.fn()
    const { getByText } = render(
      <Checkbox disabled onChange={fn}>
        测试
      </Checkbox>
    )
    const inputNode = getByText(/测试/i).previousSibling.firstChild as HTMLInputElement
    fireEvent.click(inputNode)
    expect(fn).not.toBeCalled()
  })
})
