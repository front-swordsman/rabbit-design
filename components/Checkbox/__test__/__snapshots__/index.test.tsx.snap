// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checkbox test snapshot 1`] = `"<label class=\\"snake-checkbox-item snake-checkbox-item-checked\\"><span class=\\"snake-checkbox-item-select\\"><input type=\\"checkbox\\" checked=\\"\\"><span class=\\"snake-checkbox-item-select-inner\\"></span></span><span class=\\"snake-checkbox-item-label\\">选中</span></label>"`;

exports[`checkbox test snapshot 2`] = `"<label class=\\"snake-checkbox-item\\"><span class=\\"snake-checkbox-item-select\\"><input type=\\"checkbox\\"><span class=\\"snake-checkbox-item-select-inner\\"></span></span><span class=\\"snake-checkbox-item-label\\">未选中</span></label>"`;

exports[`checkbox test snapshot 3`] = `"<label class=\\"snake-checkbox-item snake-checkbox-item-disabled\\"><span class=\\"snake-checkbox-item-select\\"><input type=\\"checkbox\\" disabled=\\"\\"><span class=\\"snake-checkbox-item-select-inner\\"></span></span><span class=\\"snake-checkbox-item-label\\">不可用</span></label>"`;

exports[`checkbox test snapshot 4`] = `"<label class=\\"snake-checkbox-item snake-checkbox-item-indeterminate\\"><span class=\\"snake-checkbox-item-select\\"><input type=\\"checkbox\\"><span class=\\"snake-checkbox-item-select-inner\\"></span></span><span class=\\"snake-checkbox-item-label\\">半选</span></label>"`;
